google/gax/__init__.py

```diff
@@ -482,3 +489,160 @@ def __next__(self):
         if self._index >= len(self._current):
             self._current = None
         return resource
+
+
+def _from_any(pb_type, any_pb):
+    if not any_pb.Is(pb_type.DESCRIPTOR):
+        raise TypeError(
+            'Could not convert {} to {}'.format(
+                any_pb.__class__.__name__, pb_type.__name__))
+
+    return pb_type.FromString(any_pb.value)
+
+
+class ResultError(GaxError):
+    """Thrown when an operation completes with an error result.
+    """
+
+    @classmethod
+    def from_operation_error(cls, operation_error):
+        """Factory: construct instance from an operation error result.
+
+        Args:
+            operation_error (google.rpc.Status): the error result of the
+                operation.
+        """
+        return cls(operation_error.message, operation_error.code)
+
+    def __init__(self, msg, code=None):
+        """Constructor.
+
+        Args:
+            msg (string): describes the error that occurred.
+            code (int, optional): the status code, which should be an enum
+                value of google.rpc.Code.
+        """
+        super(ResultError, self).__init__(msg)
+        self.code = code
+
+    def __str__(self):
+        msg = super(ResultError, self).__str__()
+        return 'ResultError({}, with code {})'.format(msg, self.code)
+
+
+class OperationFuture(object):
+    """A Future which polls a service for completion via OperationsApi."""
+
+    def __init__(self, operation, client, result_type, metadata_type,
+                 call_options=None):
+        """Constructor.
```

Comment:
The generally-accepted way is to precede with an underscore. Since this is the constructor, rather than just a particular method, the _class name itself_ can be preceded with an underscore.

As a note, that means that we are claiming that developers (in all but edge cases) will not be interacting directly with the class at all. (Calling public methods on instances that they are send by another means is okay.) Consistency is important here; if the developer has to interact with the class directly, then the constructor should probably be public.

I assume, @bjwatson, that your basic plan is that an `Operation` object will have an instance method that returns the appropriate `OperationFuture`?